Demultiplex Hiseq (and maybe others...)

guillermo-carrasco opened the pull request:
Ok, this is not finished nor fully tested, but yeah... I have one day left and I want you to continue with this if possible. 

@Galithil while I was developing I've been rebasing your PRs, so all your changes should be here, just please have a quick look so that I haven't overwritten any important change of yours. Not very used to rebasing... but should be fine looking at the commit history :)

@senthil10 most of this code should be familiar to you, there are lots of methods directly copied from scilifelab repo (_all_ of the ones in `taca/illumina/utils.py`, for example).

vezzi commented on taca/analysis/analysis.py:
```diff
@@ -130,8 +132,38 @@ def trigger_analysis(run_id):
+    """
+    if run.run_type == 'MiSeq':
+        return os.path.join(run.run_dir, 'Data', 'Intensities', 'BaseCalls', 'SampleSheet.csv')
+    else:
```
@guillermo-carrasco or who evetr wil work on this (maybe @me): I am not a big fun of "if MiSeq do something otherwise something else". There should be a choice for every sequencer type we have, or better an abstract class "run" that is specialised in MiSeq, HiSeqX, HiSeq2500 and so on.. maybe this is what happens next in the code but wnat to write it now as I am flying and this is the only work I can do given the connection :-D

vezzi commented on taca/analysis/analysis.py:
```diff
@@ -130,8 +132,38 @@ def trigger_analysis(run_id):
+
+
+def prepare_sample_sheet(run_dir, run_type, ss_origin):
+    """Prepares a non-X10 samplesheet
```
@guillermo-carrasco the same here.... this and the next can be two specialisation of an abstract method.

guillermo-carrasco replied:
ok... you wrote the methods for modifying the X-10 samplesheets, so now that it demultiplexes other stuff, needs to be take into account. But I see and share your point, we can mode this conditionals into the `Run` class.

guillermo-carrasco commented:
this is weird... tests are passing locally with the same config file used in travis :-/ any ideas?

guillermo-carrasco commented:
Yuhu! Its working for HiSeq with the last commits; i.e it started the second demultiplexing on a multi-index run

Galithil commented on taca/illumina/utils.py:
```diff
@@ -0,0 +1,242 @@
+def last_index_read(directory):
+    """Parse the number of the highest index read from the RunInfo.xml
+    """
+    read_numbers = [int(read.get("Number", 0)) for read in parsers.get_read_configuration(directory) if read.get("IsIndexedRead", "") == "Y"]
```
what is "read" ? where is it defined ?

guillermo-carrasco replied:
just continue reading the list comprehension :)

 `...int(read.get("Number", 0)) for read in parsers.get_read_configuration(directory)...`

guillermo-carrasco commented:
Ok, @vezzi , @senthil10 . This is the status of this PR and how I suggest to proceed :)

**Status**

This PR makes `TACA` to be able to demultiplex `HiSeq` runs, and eventually `MiSeq` runs as well. It has been tested, and works, for `HiSeq`, the test run (already demultiplexed) is `/srv/illumina/nosync/do_not_delete_GC/150511_D00483_0092_BC6FEHANXX`. @vezzi I fixed the bugs I told you when merging results. 

**How to proceed**
The testing run for MiSeq is `/srv/illumina/miseq_data/nosync/do_not_delete_GC/150424_M01320_0189_000000000-AEP7U`. I started `TACA` for that run (just add the `data_dir` in the config file) and it crashed, but I am pretty sure it is something related to that particular run.

You guys also need to test that this doesn't break anything for X10, sorry but I didn't have time to test :-(

I would suggest to merge this PR when you finish reviewing and proceed from there, as it closes several issues as well. I can always help with any weird bugs or merge conflicts, though it shouldn't be any since I rebased all @Galithil 's PRs into this one :)

P.S: You can delete the test runs whenever you want

senthil10 commented:
@guillermo-carrasco : Thank you for summarizing this PR :) And yes, I will definitely not hesitate to bother you regarding issues :stuck_out_tongue_winking_eye:

guillermo-carrasco commented:
what do you guys plan to do with this? @vezzi @senthil10

senthil10 commented:
@guillermo-carrasco : yes, we will do some tests and merge it :) As usual the problem is we don't have the time to do it right away :(

guillermo-carrasco commented:
:-D :+1: I was just wondering if you forgot it, hehe, tack!

vezzi commented:
@guillermo-carrasco I have integrate all your changes in the pul request that is coming

guillermo-carrasco commented:
:+1: @vezzi
